lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -166,28 +238,117 @@ llvm::Error writeArithmetic(const std::string &opName, const T *node,
     RETURN_ERR_IF_NOT(RN, "Can't unwind Tile node.");
-    proto->add_input(RN->getInput().getNode()->getName());
     reporter.insert(RN);
-  } else {
-    proto->add_input(LHS->getName());
+    LHS = RN->getInput().getNode();
   }
 
+  int axis = -1;
   const auto *RHS = node->getRHS().getNode();
   if (const TileNode *TN = llvm::dyn_cast<TileNode>(RHS)) {
+    reporter.insert(TN);
+    // unwind broadcast with tiles shape.
+    std::vector<size_t> shape;
     const ReshapeNode *RN =
-        llvm::dyn_cast<ReshapeNode>(unwindBroadcastInput(TN, reporter));
+        llvm::dyn_cast<ReshapeNode>(unwindBroadcastInput(TN, &shape, reporter));
     RETURN_ERR_IF_NOT(RN, "Can't unwind Tile node.");
-    proto->add_input(RN->getInput().getNode()->getName());
     reporter.insert(RN);
-  } else {
-    proto->add_input(RHS->getName());
+    auto input = RN->getInput();
+    RHS = input.getNode();
+
+    if (LHS->dims(0) != RHS->dims(0)) {
+      // Extract axis from available shapes, ie. input origin,
+      // reshape and target shapes.
+      llvm::ArrayRef<size_t> origin = input.dims();
+      llvm::ArrayRef<size_t> reshape = RN->getDims();
+      assert(reshape.size() == shape.size());
+      assert(shape.size() >= origin.size());
+
+      // Replace target shape dimension if it equal 1 and the correspondent
+      // reshape dimension isn't.
+      for (size_t b = 0, e = shape.size(); b < e; ++b) {
+        if (shape[b] == 1 && reshape[b] != 1) {
+          shape[b] = reshape[b];
+        }
+      }
+
+      for (int b = 0, e = shape.size() - origin.size(); b <= e && axis == -1;
+           ++b) {
+        axis = b;
+        for (size_t i = 0; i < origin.size(); ++i) {
+          if (origin[i] != shape[i + b] &&
+              (origin[i] != 1 || reshape[i + b] != 1)) {
+            axis = -1;
+            break;
+          }
+        }
+      }
+    }
   }
+
+  proto->add_input(LHS->getName());
+  proto->add_input(RHS->getName());
+
+  // Check if the shapes of LHS and RHS are different and broadcast attribute is
+  // required.
+  if (LHS->dims(0) != RHS->dims(0)) {
+    addValueAttribute(proto, "axis", axis);
+    addValueAttribute(proto, "broadcast", 1UL);
+  }
+
   return llvm::Error::success();
 }
 
+void tensorShapeFromInput(const std::string &name, TypeRef ty,
+                          ONNX_NAMESPACE::ValueInfoProto *valueProto) {
+  valueProto->set_name(name);
+  auto *type = valueProto->mutable_type();
+  auto *tensorType = type->mutable_tensor_type();
+  tensorType->set_elem_type(ONNXModelWriter::convertType(*ty));
+  auto *tensorShape = tensorType->mutable_shape();
+  const auto &dims = ty->dims();
+  for (unsigned b = 0, e = dims.size(); b < e; ++b) {
+    auto *tensorDims = tensorShape->add_dim();
+    tensorDims->set_dim_value(dims[b]);
+  }
+}
+
+/// A helper class for ordering Graph nodes in a truly post-order order.
+class GraphBackwardsPostOrderVisitor : public PostOrderVisitor {
```

Comment:
I'm confused by this -- how is this different than `GraphPreOrderVisitor`? Why is it necessary?

Reply:
There are GraphVisitor classes GraphPreOrderVisitor and GraphPortOrderVisitor, but both first visiting inputs and outputs and then only current node. In my case I need the current node got visited first and then inputs (at this time outputs already will be visited in most cases).

Reply:
Hmm, I guess I don't see how it's different than `GraphPreOrderVisitor`. You're overriding pre and post to do the same thing as `GraphPreOrderVisitor`. And visit is the same thing, visiting all Constants and then all root nodes. Can you point out where they're different?

Reply:
Yes, you are right, I have started with PostOrder having a wrong perception that PreOrder will start with root.